BlueprintUI/Sources/Layout/GridRow.swift

```diff
@@ -0,0 +1,244 @@
+            Child(width: .proportional(width), element: element)
+        }
+    }
+
+    /// The sizing and content of a `GridRow` child.
+    enum Width: Equatable {
+        /// Assign the child a fixed width equal to the payload.
+        case absolute(CGFloat)
+        /// Assign the child a proportional width of the available layout width. Note that proportional children
+        /// take proportional shares of the available layout width.
+        ///
+        /// ## Example:
+        ///     Layout width: 100
+        ///     Child A: .proportional(1)  -> 25 (100 * 1/4)
+        ///     Child B: .proportional(3) -> 75 (100 * 3/4)
+        ///
+        /// ## Example:
+        ///     Layout width: 100
+        ///     Child A: .proportional(0.25)  -> 25 (100 * 1/4)
+        ///     Child B: .proportional(0.75) -> 75 (100 * 3/4)
+        case proportional(CGFloat)
+    }
+}
+
+// MARK: - layout -
+extension GridRow {
+    struct GridRowLayout: Layout {
+        let verticalAlignment: VerticalAlignment
+        let spacing: CGFloat
+
+        static var defaultTraits: GridRow.Width {
+            .absolute(0)
+        }
+
+        func measure(in constraint: SizeConstraint, items: [(traits: Width, content: Measurable)]) -> CGSize {
+            guard items.count > 0 else {
+                return .zero
+            }
+
+            // Measure the the row to fit its tallest child.
+            let height = _frames(in: constraint, items: items).lazy.map { $0.height }.max() ?? 0
+
+            return CGSize(width: constraint.width.maximum, height: height)
+        }
+
+        func layout(size: CGSize, items: [(traits: Width, content: Measurable)]) -> [LayoutAttributes] {
+            guard items.count > 0 else {
+                return []
+            }
+
+            return _frames(in: SizeConstraint(size), isExactConstraint: true, items: items)
+                .map(LayoutAttributes.init)
+        }
+
+        /// Compute child frames in a given layout space.
+        ///
+        /// First, we measure the size of absolutely-sized children. Next, we measure the size of
+        /// proportionally-sized children, allocating them a proportion of the layout width available after spacing
+        /// and absolutely-sized children are subtracted.
+        ///
+        /// Finally, we compute the frames of children within the row from left to right, assigning a y-origin
+        /// and height which fulfill the row's vertical alignment.
+        ///
+        /// - Parameters:
+        ///   - constraint: A constraint describing the available layout space.
+        ///   - isExactConstraint: Whether or not the constraint describes the row's size exactly. During measurement,
+        ///   it does not. During layout, it does.
+        ///   - items: Children to layout.
+        private func _frames(
+            in constraint: SizeConstraint,
+            isExactConstraint: Bool = false,
```

Comment:
This is in spirit identical, but meant to avoid the additional type surface area, as it seems to be the only additional bit of information worth encoding.

That said, perhaps I should just use the existing type. Let me look into it.

Reply:
To make full use of `VectorConstraint`, I'd have to open it (and a few related extensions) up a bit. After exploring it, I decided to stick to our own type, which is scoped down to exactly the concerns we have. This type, of course, is nothing more than a `SizeConstraint` and a `Bool` at heart, but it does make the layout code marginally more legible.

Please let me know what you think!

Reply:
After playing around a bit with a few approaches, my impression is that this adds a fair bit of mostly unused complexity to `GridRow`.

That said, if we assume the average reader to already be familiar with Row, it's perhaps not a significant burden to add.

i.e.

```swift
- SizeConstraint(width: .atMost(width), height: constraint.height)
+ SizeConstraint(width: .atMost(width), height: constraint.cross.sizeConstraint)
```

```swift
- let width = content.measure(in: constraint).width
+ let width = content.measure(in: constraint.constraint(axis: .horizontal)).width
```

```swift
- let rowHeight = isExactConstraint 
-      ? constraint.height.maximum 
-      : sizes.lazy.map { $0.height }.max() ?? 0
+ let rowHeight: CGFloat
+ switch constraint.cross {
+ case .exactly(height): 
+     rowHeight = height
+ case .atMost, .unconstrained:
+    rowHeight = sizes.lazy.map { $0.height }.max() ?? 0
}
```

Reply:
OK, thanks for considering. Sometimes I find that modeling that way can make things easier to understand (see my misunderstanding of the `rowHeight` ternary below), but if it's less cumbersome to write this way, that's fine! We can ensure nothing unexpected happens with adequate test coverage.